### ISSUE TITLE
Unigame server mode has rotted a bit

### ISSUE THREAD

**Issue by joshuaauerbachwatson** [joshuaauerbachwatson/unigame#26]
The change in game token format seems to take the server by surprise.  It is posting "game token is invalid" messages to the console.

Also, the client is not seeing the "game token is invalid" message but a much more cryptic error.  Either not enough error detail is being fed back to the client or the client is not using the information.

I was confused for a while thinking the error was somehow in authentication (involving the Auth0 layer) but in fact Auth0 was happy.   The confusion probably stems from the fact that the Auth0 dialog pops up after you press the Connect button and then then after (successfully) logging into Auth0 the next thing you see is the cryptic error stemming from the "invalid" game token.

Arguably, the presence of a non-expired Auth0 token should be checked earlier and used as part of the enablement of the connect button.   The question then is how to express the need for login in a way that the user experiences login and connect as separate things.   And perhaps "connect' should be "join".

**Comment by joshuaauerbachwatson** [comment]
The Auth0 login details are in flux anyway due to issue #9 and issue #23.  I should probably do what I decided in the former issue and move the auth0 login logic back into `unigame` before trying to change the user experience of login.

**Comment by joshuaauerbachwatson** [comment]
The game token is rejected because I am using a colon character to separate the appId from the actual game token but I am expecting the server to treat the entire string as the game token.   The validator in the server is expecting at least 6 characters from the alphamerics plus hyphens and underscores.   So the colon makes the result invalid to it.

The server doesn't really need to know where the appId ends and the game token starts; the entire string works as a "group" id within the server.   So, we could "fix' this by just using underscore instead of colon.    As the composite tokens are printed to the server console, a human being can easily see where the appid ends, both because there aren't many of them and because they can't contain underscores.  

The only exposure here is that some invalid composite tokens would pass the server's validator: those whose appIds are not valid and those whose game tokens are too short.  Do I care?

It probably isn't really necessary for the server to revalidate, as the string has been assembled from components that have already been validated.   But, I think if the server is going to check, it ought to check properly and not be letting things through on a bogus basis. 

What is the server really guarding against by doing this validation?  Basically, that some user who is _not_ using the `unigame` client is sending invalid composite tokens.   But, that argues that what we should _really_ be guarding against is access to the server by people bypassing the client.

So, 

- in the immediate near term I'm changing the client to use underscore instead of colon.   This will render the server's check somewhat bogus but we don't really care very much about that.   This change doesn't require changing the server.
- the next time I'm changing the server (for example, to send better diagnostics down when websocket upgrades are rejected) I will fix the validator to be more correct (unless I immediately do the next thing)
- consider in the future restoring the "app secret" which, if properly managed, ensures that only a `unigame` client can create a websocket connection to a `unigame` server.

Note that as of today we know that only people who have valid Auth0 logins can create websocket connections to a `unigame` server.   That's almost good enough because, if necessary, we can audit who is misbehaving in other ways and cut them off.

**Comment by joshuaauerbachwatson** [comment]
> Either not enough error detail is being fed back to the client or the client is not using the information.

The server is including a body (JSON dictionary form) with an `error:` element containing the same message that is being posted to the console.  So the problem is at the client end: the client is not using this information for its own error display.   This _could_ be happening because the websocket task provided by the iOS networking framework is in the way.

**Comment by joshuaauerbachwatson** [comment]
The set of things that can be errors but the status codes are not good at conveying the actual error:
- player id failed validation
- game token failed validation
- number of players fails conversion to integer

On the assumption that the upgrade requestor is a `unigame` client, these are all _internal errors_ (should not normally occur).   They are not user errors because these things are either internally generated or validated at the client  and so being precise about them to the user is not that useful (as long as they are logged somewhere).

If the upgrade requestor is not a `unigame` client, then he is, by definition, a bad actor, because the server's API is not published.   In that case, the fact that he has an access token and can get into the server at all is already a problem, and the validators are not really part of the solution (it would be trivial for the bad actor to present valid strings rather than invalid ones).   So, we will need to use Auth0 techniques to lock this down (out of scope here). 

In short, I will just accept the limitation here and not worry about making error messages more precise.

**Comment by joshuaauerbachwatson** [comment]
> * consider in the future restoring the "app secret" which, if properly managed, ensures that only a `unigame` client can create a websocket connection to a `unigame` server.

That's one way, but probably not the best one.  As stated earlier, I need to look to Auth0 for the proper solution here.  Maybe it's already secure enough.

**Comment by joshuaauerbachwatson** [comment]
> Arguably, the presence of a non-expired Auth0 token should be checked earlier and used as part of the enablement of the connect button. The question then is how to express the need for login in a way that the user experiences login and connect as separate things. And perhaps "connect' should be "join".

Now, the `makeCommunicator` function, which is called in response to touching `Connect`, calls `CredentialStore.loginIfNeeded`, which calls the token provider iff there are not valid credentials already.   It then constructs the `ServerBasedCommunicator` using the provided token.  Proposed:

1.  There are two buttons, `Login` and `Join`.   The `Login` button is only enabled when there are _not_ valid credentials for connecting in server mode and it is also disabled if `nearbyOnly` is chosen.   The `Join` button is only enabled when all the conditions for connecting to the server are met, _including_ valid credentials.
2. The check for whether valid credentials exist is `CredentialStore.credentials`.

**Comment by joshuaauerbachwatson** [comment]
I had done most of what I planned under this issue and started testing with two players using the server.  All goes well through the socket upgrades and as long as just player lists are being exchanged.   But, the first time a game state is transmitted, the server posts: 

```
Valid message received of type 71 from player Sm9zaGll:1.  Broadcasting.
error: websocket: close 1002 (protocol error)
Closing connection for player Sm9zaGll:1
Sending lost player message for player Sm9zaGll:1
error: websocket: close 1002 (protocol error)
Closing connection for player U29sb21vbiBUZXJyZWxs:257358889
Sending lost player message for player U29sb21vbiBUZXJyZWxs:257358889
```

It will be a chore to figure out exactly where the error is first detected but a good hypothesis is that this is the first time a CBOR encoded message has been sent.   Could be something about that.

**Comment by joshuaauerbachwatson** [comment]
Looking at the debug console at messages that I didn't originate (appear to come from the network layer), it is complaining about invalid utf8.  I think this was the point at which error was detected.   The report at the server end was because the client end did a close due to protocol error.

The websocket standard allows for two kinds of messages, text and binary.   If a text message arrives with invalid utf8 the implementation MUST close the connection reporting a protocol error, so I conclude that the server sent a text frame containing invalid utf8.   A binary message does not have to be valid utf8.

The server was originally designed for chat (text) and I adapted it.  It must now send chat messages as text and game states as binary.   I will open an issue in `unigame-server` to address this.   Previously, the `anyCards` app used JSON to encode game states so text messages worked.   Now it is using CBOR.   The encoding it left to the app (`unigame` does not perform it on behalf ot he app) so only binary messages are "safe".

**Comment by joshuaauerbachwatson** [comment]
The last problem was fixed by changing the server to use binary frames for game states.   All is well now.
